### PR TITLE
[WIP] Fit api with sacred config

### PR DESCRIPTION
**Pull request by zhreshold**
An alternative implementation related to https://github.com/dmlc/gluon-cv/pull/1323 which is better(IMO) than #1323 .

- Easier to reuse common configurations(ingredients)
- Much simpler config file and command line integration during the experimenting phase
- prettier config with comments preserved!

Example

```python
from sacred import Experiment

ex = Experiment('center_net_default', ingredients=[coco_detection, train_hyperparams])

@coco_detection.config
def update_coco_detection():
    data_shape = (512, 512)  # override coco config for center_net

@coco_detection.capture
def load_dataset(root, train_splits, valid_splits, valid_skip_empty, data_shape, cleanup):
    train_dataset = COCODetection(root=os.path.join(root, 'coco'), splits=train_splits)
    val_dataset = COCODetection(root=os.path.join(root, 'coco'),
                                splits=valid_splits, skip_empty=valid_skip_empty)
    val_metric = COCODetectionMetric(val_dataset,
                                     tempfile.NamedTemporaryFile('w', delete=False).name,
                                     cleanup=cleanup,
                                     data_shape=data_shape,
                                     post_affine=get_post_transform)
    return train_dataset, val_dataset, val_metric

@ex.config
def model():
    base_network = 'dla34_deconv_dcnv2'

@set_default(ex)
class CenterNetEstimator(BaseEstimator):
    def __init__(self, config, logger=None, logdir=None):
        super(CenterNetEstimator, self).__init__(config, logger, logdir)
        # print(self._cfg)

    def _fit(self):
        pass

@ex.automain
def main(_config, _log):
    # main is the commandline entry for user w/o coding
    c = CenterNetEstimator(_config, _log)
    c.fit()
```

### One time configuration
First of all, a util function that returns the default arguments for this particular estimator is provided so that you don't have to repetitively write the docs for each argument. Though there's no excellent solution so far to inject comments into the docstring or yaml file.

### Automatic docstring generation
A decorator for injecting the default arguments into the BaseEstimator with lots of helper functions is provided. The `__doc__` will be generated by reading the default values from the config.

For example, call help(estimator) of above example will produce:

```
class CenterNetEstimator(gluoncv.pipelines.base.BaseEstimator)
 |  CenterNetEstimator(config=None, reporter=None, logdir=None)
 |
 |  CenterNet Estimator.
 |
Parameters
----------
config : str, dict
  Config used to override default configurations.
  If `str`, assume config file (.yml, .yaml) is used.
logger : logger, default is `None`.
  If not `None`, will use default logging object.
logdir : str, default is None.
  Directory for saving logs. If `None`, current working directory is used.

Default configurations:
----------
  base_network = 'dla34_deconv_dcnv2'
  coco_detection:
    cleanup = True
    data_shape = [512, 512]          # override coco config for center_net
    root = '/Users/zhiz/.mxnet/datasets'
    train_splits = 'instances_train2017'
    valid_skip_empty = False
    valid_splits = 'instances_val2017'
  train_hyperparams:
    batch_size = 32
    epochs = 3
    gpus = [0, 1, 2, 3]              # gpu individual ids, not necessarily consecutive
    num_workers = 16                 # cpu workers, the larger the more processes used
    pretrained_base = True           # use pre-trained weights from ImageNet
    resume = ''
    start_epoch = 0
    transfer_from = None
```

### Parse configurations from yaml file or command line
The `automain` entry can be used to easily accept config file or command line overrides

```
python -m gluoncv.pipelines.estimator.center_net print_config with config.yaml
```
which produce 
```
Configuration (modified, added, typechanged, doc):
  base_network = 'dla34_deconv_dcnv2'
  config_filename = 'config1.yaml'
  seed = 927480057                   # the random seed for this experiment
  coco_detection:
  1 base_network: '123'
    cleanup = True
    data_shape = [512, 512]          # override coco config for center_net
    root = '/Users/zhiz/.mxnet/datasets'
    train_splits = 'instances_train2017'
    valid_skip_empty = False
    valid_splits = 'instances_val2017'
  train_hyperparams:
    batch_size = 32
    epochs = 3
    gpus = [0, 1, 2, 3]              # gpu individual ids, not necessarily consecutive
    num_workers = 16                 # cpu workers, the larger the more processes used
    pretrained_base = True           # use pre-trained weights from ImageNet
    resume = ''
    start_epoch = 0
    transfer_from = None
```

or custom overrides
```
python -m gluoncv.pipelines.estimator.center_net print_config with 'train_hyperparams.batch_size=128'
```

```
Configuration (modified, added, typechanged, doc):
  base_network = 'dla34_deconv_dcnv2'
  config_filename = 'config1.yaml'
  seed = 927480057                   # the random seed for this experiment
  coco_detection:
  1 base_network: '123'
    cleanup = True
    data_shape = [512, 512]          # override coco config for center_net
    root = '/Users/zhiz/.mxnet/datasets'
    train_splits = 'instances_train2017'
    valid_skip_empty = False
    valid_splits = 'instances_val2017'
  train_hyperparams:
    batch_size = 128
    epochs = 3
    gpus = [0, 1, 2, 3]              # gpu individual ids, not necessarily consecutive
    num_workers = 16                 # cpu workers, the larger the more processes used
    pretrained_base = True           # use pre-trained weights from ImageNet
    resume = ''
    start_epoch = 0
    transfer_from = None
```

### Finalize config and dump to yaml file
To ensure that we don't modify the config by accident, we can call `finalize_config`, this will also save the yaml file to `logdir` for reference.
The generated 'xxx_module.xxxEstimator-06-02-2020-23-12-03.yml' file looks like this:

```yaml
base_network: dla34_deconv_dcnv2
coco_detection:
  cleanup: true
  data_shape:
  - 1024
  - 512
  root: /Users/zhiz/.mxnet/datasets
  train_splits: instances_train2017
  valid_skip_empty: false
  valid_splits: instances_val2017
seed: 772107811
train_hyperparams:
  batch_size: 32
  epochs: 3
  gpus:
  - 0
  - 1
  - 2
  - 3
  - 4
  num_workers: 32
  pretrained_base: true
  resume: ''
  start_epoch: 0
  transfer_from: null
```

### CODE REVIEW

**Comment by zhreshold** [comment]
@Jerryzcn @chongruo Please leave your comments on pro/cons about these two solutions, I personally prefer this over #1323.

File: gluoncv/pipelines/estimator/base_estimator.py [hunk 1]
```diff
@@ -0,0 +1,65 @@
...
+        cls.__doc__ += str("\n".join(_format_config(r.config, r.config_modifications).splitlines()[1:]))
+        # default config
+        cls._ex = ex
+        return cls
+    return _apply
+
+
+class BaseEstimator:
+    def __init__(self, config, logger=None, logdir=None):
+        self._ex.add_config(config)
```
Comment (Jerryzcn) [inline gluoncv/pipelines/estimator/base_estimator.py:47]: can we add some comment regarding set_defualt?

**Comment by chongruo** [comment]
I wonder if it supports multilevel inheritance? 

When users need to change some hype-parameters, is it allowed for them to create a new yaml file? thanks

**Comment by zhreshold** [comment]
@chongruo It supports multilevel inheritance and multiple way to override the configs. A new yaml is allowed, with partial/full config overriding

**Comment by zhreshold** [comment]
The doc is available here: https://sacred.readthedocs.io/en/stable/configuration.html

File: gluoncv/pipelines/estimator/base_estimator.py [hunk 2]
```diff
@@ -0,0 +1,65 @@
...
+        r = self._ex.run('_get_config', options={'--loglevel': 50})
+        self._cfg = r.config
+        self._log = logger if logger is not None else logging.getLogger(__name__)
```
Comment (Jerryzcn) [inline gluoncv/pipelines/estimator/base_estimator.py:50]: logger?

Comment (zhreshold) [reply gluoncv/pipelines/estimator/base_estimator.py:50]: instead of `logging.info` you can always use `self._log.info`

**Pull request merged**